examples/dspy-demo/utils.py

```diff
@@ -0,0 +1,124 @@
+        ```python
+        embeddings = sentence_transformer_embedding_function(["Hello, world!"])
+        ```
+    """
+
+    return embed_model.encode(sentences).tolist()
+
+
+class TidbRM(dspy.Retrieve):
+    """
+    A retrieval module that uses TiDBVectorClient to return passages for a given query.
+
+    Args:
+        tidb_vector_client (TiDBVectorClient): The TiDBVectorClient instance to use for querying TiDB.
+        embedding_function (callable): The function to convert a list of text to embeddings.
+            The embedding function should take a list of text strings as input and output a list of embeddings.
+        k (int, optional): The number of top passages to retrieve. Defaults to 3.
+
+    Returns:
+        dspy.Prediction: An object containing the retrieved passages.
+
+    Examples:
+        Below is a code snippet that shows how to use this as the default retriever:
+        use OpenAI
+        ```python
+        llm = dspy.OpenAI(model="gpt-3.5-turbo")
+        retriever_model = TidbRM(
+            tidb_vector_client=tidb_vector_client,
+            embedding_function=sentence_transformer_embedding_function
+        )
+        dspy.settings.configure(rm=retriever_model)
+        ```
+
+        use Ollama
+        ```python
+        llm = dspy.OllamaLocal(model="llama3:8b")
+        retriever_model = TidbRM(
+            tidb_vector_client=tidb_vector_client,
+            embedding_function=llm
+        )
+
+    """
+
+    def __init__(self, tidb_vector_client: TiDBVectorClient, embedding_function: Optional[callable] = None, k: int = 3):
+        super().__init__(k)
+        self.tidb_vector_client = tidb_vector_client
+        self.embedding_function = embedding_function
+        self.top_k = k
+
+    def forward(self, query_or_queries: Union[str, List[str]], k: Optional[int] = None, **kwargs) -> dspy.Prediction:
+        """
+        Retrieve passages for the given query.
+
+        Args:
+            query_or_queries (Union[str, List[str]]): The query or queries for which to retrieve passages.
+            k (Optional[int]): The number of top passages to retrieve. Defaults to 3.
+
+        Returns:
+            dspy.Prediction: An object containing the retrieved passages.
+
+        Examples:
+            Below is a code snippet that shows how to use this function:
+            ```python
+            passages = self.retrieve("Hello, world!")
+            ```
+        """
+        query_embeddings = self.embedding_function(query_or_queries)
+        k = k or self.top_k
+        tidb_vector_res = self.tidb_vector_client.query(query_vector=query_embeddings, k=k)
+        passages_scores = {}
+        for res in tidb_vector_res:
+            res.metadata = dotdict(res.metadata)
+            passages_scores[res.document] = res.distance
+        sorted_passages = sorted(passages_scores.items(), key=lambda x: x[1], reverse=True)
+        k_sorted_passages = sorted_passages[:k]
```

Comment:
Do we need to select the top K items again after invoking `self.tidb_vector_client.query(query_vector=query_embeddings, k=k)`?

Reply:
Not necessary, the parameter k in the code snippet has no effect. 
I will remove it.